h2o-algos/src/main/java/hex/glm/GLM.java

```diff
@@ -1572,25 +1578,62 @@ private void fitIRLSM_multinomial(Solver s) {
+                      ? new MoreThuente(_state.gslvrMultinomial(c), _state.betaMultinomialFull(c, beta), _state.ginfoMultinomial(c))
+                      : new SimpleBacktrackingLS(_state.gslvrMultinomial(c), _state.betaMultinomialFull(c, beta), _state.l1pen());
+            else if (_parms._remove_collinear_columns && _state._iter > 0)
+              ls = (_state.l1pen() == 0)  // after first iteration over all classes, state._beta, ginfo.gradient are all smaller size
+                      ? new MoreThuente(_state.gslvrMultinomial(c), _state.betaMultinomialFull(c, beta), _state.ginfoMultinomialRCC(c))
+                      : new SimpleBacktrackingLS(_state.gslvrMultinomial(c), _state.betaMultinomialFull(c, beta), _state.l1pen());
+            else
+              ls = (_state.l1pen() == 0)
                     ? new MoreThuente(_state.gslvrMultinomial(c), _state.betaMultinomial(c, beta), _state.ginfoMultinomial(c))
                     : new SimpleBacktrackingLS(_state.gslvrMultinomial(c), _state.betaMultinomial(c, beta), _state.l1pen());
 
             long t1 = System.currentTimeMillis();
-            new GLMMultinomialUpdate(_state.activeDataMultinomial(), _job._key, beta, c).doAll(_state.activeDataMultinomial()._adaptedFrame);
+            // GLMMultinomialUpdate needs to take beta that contains active columns described in _state.activeDataMultinomial()
+            if (_parms._remove_collinear_columns && _state.activeDataMultinomial()._activeCols != null && 
+            _betaLenPerClass != _state.activeDataMultinomial().activeCols().length) { // beta full length, need short beta
+              double[] shortBeta = _state.shrinkFullArray(beta);
+              new GLMMultinomialUpdate(_state.activeDataMultinomial(), _job._key, shortBeta,
+                              c).doAll(_state.activeDataMultinomial()._adaptedFrame);        
+            } else {
+              new GLMMultinomialUpdate(_state.activeDataMultinomial(), _job._key, beta,
+                      c).doAll(_state.activeDataMultinomial()._adaptedFrame);
+            }
             long t2 = System.currentTimeMillis();
-            ComputationState.GramXY gram = _state.computeGram(ls.getX(), s);
-            long t3 = System.currentTimeMillis();
-            double[] betaCnd = ADMM_solve(gram.gram, gram.xy);
+            ComputationState.GramXY gram;
+            if (_parms._remove_collinear_columns && _state._iter > 0)
+              gram = _state.computeGramRCC(ls.getX(), s);
+            else
+              gram = _state.computeGram(ls.getX(), s);
 
+            long t3 = System.currentTimeMillis();
+            double[] betaCnd;
+            if (_parms._remove_collinear_columns && _state._iter > 0)
+              betaCnd = ADMM_solve(gram.gram, gram.xy);
+            else 
+              betaCnd = ADMM_solve(gram.gram, gram.xy);  // rcc=true, remove inactive col from _state beta, ginfo
```

Comment:
This looks like an unnecessary condition or bug:
```suggestion
            betaCnd = ADMM_solve(gram.gram, gram.xy);
```

Reply:
Tomas:  The two are different.  One is calling computeGramRCC and one is calling computeGram.

Reply:
You are right.  It is the same!